### ISSUE TITLE
The insertContent command adds unnecessary blank paragraphs with HTML as value

### ISSUE THREAD

**Issue by rfgamaral** [ueberdosis/tiptap#2720]
### What’s the bug you are facing?

The `insertCommand` seems to be adding unnecessary blank paragraphs to the editor when newlines are found between HTML tags.

### Which browser was this experienced in? Are any special extensions installed?

n/a

### How can we reproduce the bug on our side?

Take the following HTML as an example:

```js
const htmlInput= `<h1><a href="https://tiptap.dev/">Tiptap</a></h1>
<p><strong>Hello World</strong></p>`
```

And then call `editor.commands.insertCommand(htmlInput)`. This is what gets rendered by Tiptap:

```html
<div
    translate="no"
    role="textbox"
    aria-readonly="false"
    aria-multiline="true"
    class="ProseMirror"
    tabindex="0"
    contenteditable="true"
>
    <p><br class="ProseMirror-trailingBreak" /></p>
    <h1>
        <a target="_blank" rel="noopener noreferrer nofollow" href="https://tiptap.dev">Tiptap</a>
    </h1>
    <p></p>
    <p><strong>Hello World</strong></p>
</div>
```

As you can see, there's one paragraph before and after the header that shouldn't be there.

### What did you expect to happen?

Based on the example above, Tiptpa should've rendered this:

```html
<div
    translate="no"
    role="textbox"
    aria-readonly="false"
    aria-multiline="true"
    class="ProseMirror"
    tabindex="0"
    contenteditable="true"
>
    <h1>
        <a target="_blank" rel="noopener noreferrer nofollow" href="https://tiptap.dev">Tiptap</a>
    </h1>
    <p><strong>Hello World</strong></p>
</div>
```

### Did you update your dependencies?

- [X] Yes, I’ve updated my dependencies to use the latest version of all packages.

### Are you sponsoring us?

- [X] Yes, I’m a sponsor. 💖 _(on behalf of Doist)_

**Comment by rfgamaral** [comment]
@bdbch This issue still exists, do you think this could be reopened, and possibly some attention?

**Comment by bdbch** [comment]
bump

**Comment by evanshe** [comment]
The same with u, instead , i call insertcontent, value is the json
```
inserContent({
  type: 'xxx',
content: []
})
```

**Comment by ondrejsevcik** [comment]
@determinationlove I don't think that comment like this is appropriate.

> Errr... Can't believe this bug has been reported for more than a year without being fixed?

This is open source project that you can use for free. Unless you're paying for it, you can only appreciate the work of the maintainers that they put into it. And also report bugs and issues, but always in a polite way! They are not working for you.

**Comment by sakshi-klaarhq** [comment]
Hello, is there a way to solve this problem ?
I have encountered it as well in the taskItems extension

**Comment by bdbch** [comment]
@sakshi-klaarhq did you experience it by using insertContentAt manually to insert a taskList or was your issue caused by the taskList extension itself?

**Comment by rajatkulkarni95** [comment]
Was running into the same issue, but adding `parseOptions` to not `preserveWhitespace` made it work

```
  editor.chain()
            .focus()
            .deleteRange(range)
            .insertContent(res.content, {
              parseOptions: {
                preserveWhitespace: false,
              },
            })
            .run();
        },
```


Here's a before after video comparison

https://github.com/ueberdosis/tiptap/assets/57321156/27dd8463-2204-4530-85aa-9f7183bbd3d7

**Comment by rdewolff** [comment]
This bug is still here.

Inserting html code with a list, tiptap adds paragraph tags <p> on every line. 

For example:

When using `editor.chain().focus().insertContentAt({ x, y }, text).run()`

With:

```
const text = `<h1>Liste des Courses</h1>
    <ul>
        <li>Un gâteau à la crème</li>
        <li>Un gâteau au chocolat</li>
        <li>Des carottes</li>
        <li>Des tomates</li>
    </ul>`
```

The HTML in TipTap editor is the following:

```
<p></p><h1>Liste des Courses!</h1><p> </p><ul><li><p> </p></li><li><p>Un gâteau à la crème</p></li><li><p> </p></li><li><p>Un gâteau au chocolat</p></li><li><p> </p></li><li><p>Des carottes</p></li><li><p> </p></li><li><p>Des tomates</p></li><li><p> </p></li></ul>
```

Anyone found a workaround for this bug?

**Comment by rdewolff** [comment]
This is a codesandbox repo where you can see and reproduce the bug in action:

https://codesandbox.io/p/devbox/editor-forked-p376l7?file=%2Fsrc%2FApp.js

1. load it
2. hit the "ADD HTML LIST" button on the top menubar

See the resulted list broken 😭

**Comment by rdewolff** [comment]
I have been able to fix it:

Just remove any space between your HTML tags and the layout will be fine.

**Comment by vucuong1820** [comment]
> I have been able to fix it:
> 
> Just remove any space between your HTML tags and the layout will be fine.

Could you show me your solution in code?  @rdewolff

**Comment by rdewolff** [comment]
Sure, I just use regexp to cleanup the HTML code, as follow:

```
  function cleanHTML(htmlString) {
    return htmlString.replace(/\s*(<[^>]+>)\s*/g, '$1')
  }
```

Let me know if that works for you 😉

**Comment by gkubisa** [comment]
> Sure, I just use regexp to cleanup the HTML code, as follow:
> 
> ```ts
> export const cleanHTML = (htmlString) => {
>   // This will remove all the spaces between HTML tags and words. It is too aggressive.
>   // return htmlString.replace(/\s*(<[^>]+>)\s*/g, '$1')
>   // This pattern specifically targets spaces that occur between a closing tag (>) and an opening tag (<), effectively removing only those spaces.
>   return htmlString.replace(/>\s+</g, '><')
> }
> ```
> 
> Let me know if that works for you 😉

In my opinion this is the correct solution. :+1: 

TipTap should not make assumptions about which white space to remove because there'll be always edge cases.

We should actually revert https://github.com/ueberdosis/tiptap/pull/4767 because it breaks the expected functionality within `<pre>`, for example:

```html
<pre>
<span>a</span>
<span>b</span>
</pre>
```

![image](https://github.com/user-attachments/assets/6fee3feb-18d8-4585-a270-a515ffecf0eb)

```html
<pre>
<span>a</span>
 <span>b</span>
</pre>
```

![image](https://github.com/user-attachments/assets/1cfbcbaa-9bfc-497a-bfdf-de29015a0353)

```html
<pre>
<span>a</span>
  <span>b</span>
</pre>
```

![image](https://github.com/user-attachments/assets/985f7982-b46c-4316-b071-c2325970fe41)

For now I'll deal with the issue above in my own code by cleaning the content of `<pre>` before passing it to TipTap, however, that `removeWhitespaces` function is really doing more harm than good and should be removed.

**Comment by Beslinda** [comment]
For ListItems you can do the following and it will not add the additional paragraph.

```
import ListItem from '@tiptap/extension-list-item';

const CustomListItem = ListItem.extend({
  content: 'text*'
});
```

and add the CustomListItem on the extensions `extensions: [......, CustomListItem]`

**Comment by srcrip** [comment]
For anyone coming from google this seems to still be a problem. I'm able to avoid it by using:

```
          {
            parseOptions: {
              preserveWhitespace: false,
            },
          }
```

as suggested above. But it'd be nice if it got fixed upstream.

**Comment by rfgamaral** [comment]
The original issue I posted is no longer an issue for us, and I can't personally reproduce it. Also, it seems to me that there are multiple (albeit similar) issues being discussed here, some with workarounds. For those reasons, I'm closing this myself.

Feel free to open new and more specific issues for any problems you might still be having with the current behaviour.

**Comment by srcrip** [comment]
I am actually still seeing these problems now when I try to insert over a selection of the document.

**Comment by SawkaDev** [comment]
Bump, see the same thing

**Comment by Rafik-Belkadi-Reccap** [comment]
> Sure, I just use regexp to cleanup the HTML code, as follow:
> 
> export const cleanHTML = (htmlString) => {
>   // This will remove all the spaces between HTML tags and words. It is too aggressive.
>   // return htmlString.replace(/\s*(<[^>]+>)\s*/g, '$1')
>   // This pattern specifically targets spaces that occur between a closing tag (>) and an opening tag (<), effectively removing only those spaces.
>   return htmlString.replace(/>\s+</g, '><')
> }
> Let me know if that works for you 😉



Added a slightly better solution, yours was a bit too aggressive ^^

```ts
private cleanHTML = (htmlString: string): string => {
        // This will remove all the spaces between HTML tags and words. It is too aggressive.
        // return htmlString.replace(/\s*(<[^>]+>)\s*/g, '$1')
        // First, remove multiple spaces between tags
        const noMultiSpacesBetweenTags = htmlString.replace(/>\s+</g, '><');
        // Then, replace multiple spaces with a single space in text content
        return noMultiSpacesBetweenTags.replace(/\s{2,}/g, ' ');
    };```

**Comment by bdbch** [comment]
I'm reverting the fix made it #4767 as it breaks a lot of situations where whitespaces should be kept.

ProseMirror should handle whitespace handling already.
